pkg/kubectl/cmd/util/diff/diffoptions.go

```
@@ -0,0 +1,183 @@
+
+		o.DiffBufferOriginal, err = runtime.Encode(codec, targetObj)
+		return err
+	})
+
+	return err
+}
+
+func deleteEmptyAnnotations(obj runtime.Object) error {
+	if unstructured, ok := obj.(runtime.Unstructured); ok {
+		if tarObj, ok := unstructured.UnstructuredContent()["metadata"].(map[string]interface{}); ok {
+			delete(tarObj, "annotations")
+			return nil
+		}
+	}
+	return fmt.Errorf("delete empty annotations failed")
+}
+
+// Validate validates all the required options for apply diff-last-applied.
+func (o *Options) Validate(f cmdutil.Factory, cmd *cobra.Command) error {
+	return nil
+}
+
+// RunDiffLastApplied implements all the necessary functionality for apply diff-last-applied.
+func (o *Options) RunDiffLastApplied(f cmdutil.Factory, cmd *cobra.Command) error {
+	var err error
+	o.DiffBufferModified, err = FormatConvert(o.DiffBufferModified, o.Output)
+	if err != nil {
+		return err
+	}
+	o.DiffBufferOriginal, err = FormatConvert(o.DiffBufferOriginal, o.Output)
+	if err != nil {
+		return err
+	}
+	diff := cmdutil.NewDefaultCmdTool(cmdutil.DiffCmd, f.DiffEnvs())
+
+	tmpFileSlice := []cmdutil.TempFile{
+		//be careful with the order of DiffBufferModified and DiffBufferOriginal
+		{Prefix: "diff-last-applied-original-", Suffix: "." + o.Output, Buffer: bytes.NewReader(o.DiffBufferOriginal)},
+		{Prefix: "diff-last-applied-modified-", Suffix: "." + o.Output, Buffer: bytes.NewReader(o.DiffBufferModified)},
+	}
+	_, err = diff.LaunchTempFiles(tmpFileSlice...)
+
+	return err
+}
+
+// FormatConvert accept a JSON-encoded byte array and a convert type
+// returned a JSON or YAML formatted byte array and an error
+func FormatConvert(input []byte, o string) ([]byte, error) {
+	switch o {
+	case "json":
+		var prettyJSON bytes.Buffer
+		err := json.Indent(&prettyJSON, input, "", "  ")
+		if err != nil {
+			return nil, err
+		}
+		return prettyJSON.Bytes(), nil
+	case "yaml":
+		return yaml.JSONToYAML(input)
+	}
+	return nil, fmt.Errorf("unexpected -o output mode")
```

Personally I would panic at the end of the function rather than silently returning nil.